HTTP API: allow connections to be listed and closed by username

NuwanSameera opened the pull request:
This feature was added to RabbitMQ management plugin. New HTTP API added to delete connection by username. 

Following is the API 
`/api/connections/username/admin-test`

Current implementation provides delete connection by connection name. Connection name doesn't have known parameters. User name is a known parameter. This feature is important to delete connection programmatically in SSL certificate login enable server. In this case user name contains client CN key of the certificate. 

If RabbitMQ uses custom authentication backend, this feature is very useful. We can implement logic to find valid CN key and we can remove connection if that logic says username is invalid. With this implementation we can block reconnection.

See:
* #5275 
* #5251

lukebakken commented on deps/rabbit/src/rabbit_connection_tracking.erl:
```diff
@@ -47,10 +47,11 @@
          clear_tracked_connection_tables_for_this_node/0,
 
-         list/0, list/1, list_on_node/1, list_on_node/2, list_of_user/1,
+         list/0, list/1,list_on_node/1, list_on_node/2, list_of_user/1,
```
Please revert this whitespace change.

lukebakken commented on deps/rabbitmq_management/src/rabbit_mgmt_wm_connection_user_name.erl:
```diff
@@ -0,0 +1,87 @@
+    end,
+    {true, ReqData, Context}.
+
+
```
Remove the extra newline.

lukebakken reviewed (commented):
I fixed the whitespace and one compilation warning. This API is different (and better!) than what was previously described.

lukebakken reviewed (commented):
Instead of only supporting `DELETE`, we should support `GET` as well to list connections by username.

I'm not 100% convinced that `/connections/username` is the best REST resource for this. @michaelklishin thoughts? How about this

`/connections?username=FOO`

I think the version with the query parameter makes the most sense. I'll see if we do something similar elsewhere in the API.

Good discussion here: https://stackoverflow.com/questions/34244627/getting-a-resource-using-another-resources-id-naming-in-rest-api

@NuwanSameera please note the API docs will have to be updated as well - https://github.com/rabbitmq/rabbitmq-server/blob/master/deps/rabbitmq_management/priv/www/api/index.html

NuwanSameera commented:
@lukebakken I updated API docs for DELETE API. 

Also I enabled and check for GET request for same code. But it gave 500 error.

lukebakken commented on deps/rabbit/src/rabbit_connection_tracking.erl:
```diff
@@ -365,6 +386,8 @@ list() ->
       end, [], rabbit_nodes:all_running()).
 
+
+
```
Please remove these two blank lines.

NuwanSameera replied:
Done

lukebakken reviewed (changes requested):
Hi @NuwanSameera -

> Also I enabled and check for GET request for same code. But it gave 500 error.

I don't see changes in your pull request for enabling `GET`. Did you push them to your fork at https://github.com/NuwanSameera/rabbitmq-server.git ?

NuwanSameera commented:
@lukebakken Enabled GET method to get all connections by username.
API documentation also updated.

michaelklishin commented:
`/connections/by/username/{username}` or `/connections/filtered/username/{username}` would be fine with me.

An overly generic endpoint that allows filtering by every potential argument will quickly
grow large and complex. On top of that, query string values are rarely validated and supported
paths by definition are.

michaelklishin commented:
@NuwanSameera thank you for considering to contribute. This looks like a promising start.
It lacks a CLI command but given all core API elements in place, we can easily add one on your behalf.

Note that we are about to switch from Mnesia tables to a node-local ETS one for connection tracking in https://github.com/rabbitmq/rabbitmq-server/pull/5301. This is significantly
more efficient and helps with the footprint in high churn scenarios. That's for 3.11 only.

That means that this PR will have to change to adapt and for 3.10, we'd have to submit
a version that looks like what we have today.

NuwanSameera commented:
Have any additional work need to do from myside.

mkuratczyk commented:
Thanks a lot for taking the time to work on a feature you need!

CLI support already exists: https://github.com/rabbitmq/rabbitmq-server/pull/2771. There's some code that could be shared:
* PR2771 relies on `rabbit_connection_tracking:list_of_user/1`, so I think `list_by_username/1` is redundant
* PR2771 added `rabbit_networking:close_all_user_connections/2`, which I think could be used in this PR

NuwanSameera commented:
@mkuratczyk I updated code with suggested methods. List all connections method is working properly. 
But delete all connections method didn't work for me. 

Following method every time return empty list. 
https://github.com/NuwanSameera/rabbitmq-server/blob/ccc4f69471cf073eeaedf268f296387396edc333/deps/rabbit/src/rabbit_networking.erl#L411 

So 

https://github.com/NuwanSameera/rabbitmq-server/blob/ccc4f69471cf073eeaedf268f296387396edc333/deps/rabbit/src/rabbit_networking.erl#L477

condition false every time. Any mistake in my code ?

NuwanSameera commented:
@mkuratczyk what about MQTT connections ?

michaelklishin commented:
MQTT connections likely use a separate set of functions that lists them. Querying tables
directly worked around this.

I'd wait for https://github.com/rabbitmq/rabbitmq-server/pull/5301 to be finished, then use whatever newer version of the connection tracking API we would end up with, then backport
and adapt to the API `3.10.x` has/will have.

NuwanSameera commented:
I updated PR with changing delete method. It is tested with MQTT connections. Also removed duplicate list connection method.

Is it able to get official docker image with these APIs ?

mkuratczyk commented:
The official image (`rabbitmq`) will be built once we release a new RabbitMQ version with the changes merged. For PRs, we generally build test images (https://hub.docker.com/r/pivotalrabbitmq/rabbitmq/tags), but since your PR comes from a fork, this process wasn't triggered. You can try `make package-generic-unix && make docker-image` to build an image locally.

michaelklishin commented:
AFAIK all PRs related to connection tracking are in. @NuwanSameera can you please rebase and we will QA? Thank you.

NuwanSameera commented:
@michaelklishin 
`git rebase --onto master v3.10.6 feature-delete-connection-by-username`
Is this correct command to rebase ?

michaelklishin commented:
@NuwanSameera simply pull `master, then switch back to your branch, `git rebase master` and force-push to your fork using something like `git push -u {name of your remote} HEAD -f`.

NuwanSameera commented:
@michaelklishin pushed changes to my fork branch. But it is not compile. 

Following error getting when executing `make run-broker`
```
Generated escript escript/rabbitmqctl with MIX_ENV=dev
make[2]: Leaving directory '/home/STL-NUWAN/rabbitmq-server/deps/rabbitmq_cli'
make[2]: Entering directory '/home/STL-NUWAN/rabbitmq-server/deps/amqp10_common'
make[2]: Leaving directory '/home/STL-NUWAN/rabbitmq-server/deps/amqp10_common'
make[2]: *** No rule to make target 'include/feature_flags.hrl', needed by 'src/rabbit_core_ff.erl'.  Stop.
make[1]: *** [../../erlang.mk:5197: app] Error 2
make[1]: Leaving directory '/home/STL-NUWAN/rabbitmq-server/deps/rabbit'
make: *** [erlang.mk:4512: deps] Error 2

```

michaelklishin commented:
There were quite a few changes in master in the last week or so. You have to run

```
gmake distclean
```

from repo root.

michaelklishin commented:
```
git clone https://github.com/NuwanSameera/rabbitmq-server.git
cd rabbitmq-server
git checkout feature-delete-connection-by-username
cd deps/rabbit
gmake
```

does successfully build for me. So does running a node from source using

```
gmake run-broker PLUGINS="rabbitmq_management"
```

I use Bazel exclusively these days but it's clearly an issue you'd run into with Make when you switch branches or rebase as master moves quickly.

You can try

```
gmake -C deps/rabbit/apps/rabbitmq_prelaunch distclean
gmake -C deps/rabbit_common distclean
gmake -C deps/rabbit distclean

gmake -C deps/rabbit/apps/rabbitmq_prelaunch
gmake -C deps/rabbit_common
gmake -C deps/rabbit
```

to clean all the core bits that have changed recently and recompile them.

NuwanSameera commented:
@michaelklishin work after following your suggested steps. I tested list and close connections with latest code. It is work fine.

Thank you.

lukebakken commented:
I am back from PTO and will find time to re-review this PR. Thanks for your patience @NuwanSameera

lukebakken commented:
> An overly generic endpoint that allows filtering by every potential argument will quickly
grow large and complex. On top of that, query string values are rarely validated and supported
paths by definition are.

I think a query parameter is the correct solution here. We use them in other places in the management API code. I'll make the necessary modifications and will create a branch with the changes.

Stay tuned.

lukebakken commented:
@NuwanSameera @michaelklishin @mkuratczyk please see the following PR which shows how I think this should be implemented, by using a query parameter:

https://github.com/NuwanSameera/rabbitmq-server/pull/4

NuwanSameera commented:
@lukebakken 

I start broker with latest commit in [feature-delete-connection-by-username](https://github.com/NuwanSameera/rabbitmq-server/tree/feature-delete-connection-by-username) . Get method is working, but DELETE method gave `405 Method Not Allowed` error

lukebakken commented:
@NuwanSameera - It helps us if you show the exact command you run and its output. There is something up with the changes I made and I'm investigating now.

lukebakken commented:
@NuwanSameera @michaelklishin @mkuratczyk this PR is ready for a re-test. Here's what I see when `DELETE`-ing a connection now:

```
$ curl --ipv4 --verbose --user 'guest:guest' --request DELETE localhost:15672/api/connections/username/guest
*   Trying 127.0.0.1:15672...
* Connected to localhost (127.0.0.1) port 15672 (#0)
* Server auth using Basic with user 'guest'
> DELETE /api/connections/username/guest HTTP/1.1
> Host: localhost:15672
> Authorization: Basic Z3Vlc3Q6Z3Vlc3Q=                                                                                                                                                                                                                                         > User-Agent: curl/7.84.0
> Accept: */*
>                                                                                                                                                                                                                                                                               * Mark bundle as not supporting multiuse
< HTTP/1.1 204 No Content
< content-security-policy: script-src 'self' 'unsafe-eval' 'unsafe-inline'; object-src 'self'
< date: Fri, 05 Aug 2022 13:17:28 GMT
< server: Cowboy
< vary: accept, accept-encoding, origin
<
* Connection #0 to host localhost left intact
```

Program that had the connection open:

```
06:17:28 INFO:process_data_events
06:17:29 INFO:Aborting transport connection: state=1; <socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 56712), raddr=('127.0.0.1', 5672)>
06:17:29 INFO:_AsyncTransportBase._initate_abort(): Initiating abrupt asynchronous transport shutdown: state=1; error=None; <socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 56712), raddr=('127.0.0.1', 5672)>
06:17:29 INFO:Deactivating transport: state=1; <socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 56712), raddr=('127.0.0.1', 5672)>
06:17:29 INFO:AMQP stack terminated, failed to connect, or aborted: opened=True, error-arg=None; pending-error=ConnectionClosedByBroker: (320) 'CONNECTION_FORCED - Closed via management plugin'
06:17:29 INFO:Stack terminated due to ConnectionClosedByBroker: (320) 'CONNECTION_FORCED - Closed via management plugin'
06:17:29 INFO:Closing transport socket and unlinking: state=3; <socket.socket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 56712), raddr=('127.0.0.1', 5672)>
06:17:29 ERROR:Unexpected connection close detected: ConnectionClosedByBroker: (320) 'CONNECTION_FORCED - Closed via management plugin'
Traceback (most recent call last):
  File "/home/lbakken/development/lukebakken/rabbitmq/pika/./1257/repro.py", line 39, in <module>
    connection.process_data_events(time_limit=5)
  File "/home/lbakken/development/lukebakken/rabbitmq/pika/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 842, in process_data_events
    self._flush_output(timer.is_ready, common_terminator)
  File "/home/lbakken/development/lukebakken/rabbitmq/pika/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 523, in _flush_output
    raise self._closed_result.value.error
pika.exceptions.ConnectionClosedByBroker: (320, 'CONNECTION_FORCED - Closed via management plugin')
```

NuwanSameera commented:
@lukebakken 
What is the API need to use
`localhost:15672/api/connections/username/guest` or `localhost:15672/api/connections?username=guest`

lukebakken commented:
@NuwanSameera we decided to keep your original API -

```
localhost:15672/api/connections/username/guest
```

michaelklishin commented:
I will take a look first thing tomorrow morning

NuwanSameera commented:
Have any estimated date to release this feature ?

michaelklishin commented:
It will ship in 3.11.0. We do not make any ETA promises.

NuwanSameera commented:
In the release note of `RabbitMQ 3.11.0-rc.1` mentioned this feature as

`New endpoint for listing connections of a specific user`. I think better if it change as `New endpoint for listing and closing connections of a specific user`

michaelklishin commented:
This shipped with a bug where for un(der)privileged users these new endpoints failed with a 500, addressed in https://github.com/rabbitmq/rabbitmq-server/pull/8483.

It now also has some test coverage in https://github.com/michaelklishin/rabbit-hole/pull/271.
